peas/tests/serial_handlers/protocol_arduino_simulator.py

```diff
@@ -0,0 +1,422 @@
+                break
+            if timeout_obj.expired():
+                break
+        response = bytes(response)
+        self.logger.info('FakeArduinoSerialHandler.read({}) -> {!r}', size, response)
+        return response
+
+    @property
+    def out_waiting(self):
+        """The number of bytes in the output buffer."""
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        # Not an accurate count because the elements of self.relay_queue are arrays, not individual
+        # bytes.
+        return self.relay_queue.qsize()
+
+    def reset_output_buffer(self):
+        """Clear output buffer.
+
+        Aborts the current output, discarding all that is in the output buffer.
+        """
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        _drain_queue(self.relay_queue)
+
+    def flush(self):
+        """Write the buffered data to the output device.
+
+        We interpret that here as waiting until the simulator has taken all of the
+        entries from the queue.
+        """
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        while not self.relay_queue.empty():
+            time.sleep(0.01)
+
+    def write(self, data):
+        """Write the bytes data to the port.
+
+        Args:
+            data: The data to write (bytes or bytearray instance).
+
+        Returns:
+            Number of bytes written.
+
+        Raises:
+            SerialTimeoutException: In case a write timeout is configured for
+                the port and the time is exceeded.
+        """
+        if not isinstance(data, (bytes, bytearray)):
+            raise ValueError("write takes bytes")
+        data = bytes(data)  # Make sure it can't change.
+        self.logger.info('FakeArduinoSerialHandler.write({!r})', data)
+        count = len(data)
+        try:
+            self.relay_queue.put(data, block=True, timeout=self.write_timeout)
+            return len(data)
+        except queue.Full:
+            # This exception is "lossy" in that the caller can't tell how much was written.
+            raise serialutil.writeTimeoutError
+
+    # --------------------------------------------------------------------------
+
+    @property
+    def is_config_ok(self):
+        """Does the caller ask for the correct serial device config?"""
+        # The default Arduino data, parity and stop bits are: 8 data bits, no parity, one stop bit.
+        return (self.baudrate == 9600 and self.bytesize == serialutil.EIGHTBITS and
+                self.parity == serialutil.PARITY_NONE and not self.rtscts and not self.dsrdtr)
+
+    def _read1(self, timeout_obj):
+        """Read 1 byte of input, of type bytes."""
+        if not self.is_open:
+            raise serialutil.portNotOpenError
+        if not self.json_bytes:
+            try:
+                entry = self.json_queue.get(block=True, timeout=timeout_obj.time_left())
+                assert isinstance(entry, bytes)
```

Comment:
I would catch this assertion also.

Reply:
Where?

Reply:
Probably not necessary for the test I suppose. Same thing as above.